source/server/drain_manager_impl.cc

```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
     return false;
   }
 
-  // We use the tick time as in increasing chance that we shutdown connections.
-  return static_cast<uint64_t>(drain_time_completed_.load()) >
-         (server_.random().random() % server_.options().drainTime().count());
-}
-
-void DrainManagerImpl::drainSequenceTick() {
-  ENVOY_LOG(trace, "drain tick #{}", drain_time_completed_.load());
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
+  const auto remaining_time = std::chrono::duration_cast<std::chrono::seconds>(
+      drain_deadline_ - server_.dispatcher().timeSource().monotonicTime());
+  const auto elapsed_time = server_.options().drainTime() - remaining_time;
 
-  if (drain_time_completed_.load() < server_.options().drainTime().count()) {
-    drain_tick_timer_->enableTimer(std::chrono::milliseconds(1000));
-  } else if (drain_sequence_completion_) {
-    drain_sequence_completion_();
-  }
+  // P(shut down connections) = elapsed_drain_time / overall_drain_timeout
+  return static_cast<uint64_t>(elapsed_time.count()) >
+         (server_.random().random() % server_.options().drainTime().count());
 }
 
-void DrainManagerImpl::startDrainSequence(std::function<void()> completion) {
-  drain_sequence_completion_ = completion;
+void DrainManagerImpl::startDrainSequence(std::function<void()> drain_complete_cb) {
+  drain_complete_cb_ = drain_complete_cb;
```

Comment:
why do we need to hold this callback in the DrainManagerImpl? Isn't it enough to hold it in the timer?

Reply:
I wasn't really sure about the lifetime parameters of `std::function`s but that's fine by me.